Lib/typing.py

```diff
@@ -823,6 +874,117 @@ def __init__(self, name, *constraints, bound=None,
             self.__module__ = def_mod
 
 
+class TypeVarTuple(_Final, _Immutable, _root=True):
+    """Type variable tuple.
+
+    Usage:
+
+      Ts = TypeVarTuple('Ts')  # Can be given any name
+
+    Just as a TypeVar (type variable) is a placeholder for a single type,
+    a TypeVarTuple is a placeholder for an *arbitrary* number of types. For
+    example, if we define a generic class using a TypeVarTuple:
+
+      class C(Generic[*Ts]): ...
+
+    Then we can parameterize that class with an arbitrary number of type
+    arguments:
+
+      C[int]       # Fine
+      C[int, str]  # Also fine
+      C[()]        # Even this is fine
+
+    For more details, see PEP 646.
+    """
+
+    def __init__(self, name):
+        self._name = name
+        self._unpacked = UnpackedTypeVarTuple(name, self)
+
+    def __iter__(self):
+        yield self._unpacked
+
+    def __repr__(self):
+        return self._name
+
+
+class UnpackedTypeVarTuple(_Final, _Immutable, _root=True):
```

Comment:
This could also be implemented as `_GenericAlias(Unpack, (Ts,))`, although I acknowledge that makes some of the below more complicated.

Reply:
Wasn't too bad :)